File: c/tree/recover-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *prev, *t2,*t1;

void inorder(struct TreeNode *root){
    if(root == NULL)
        return;

    inorder(root->left);

    if(prev && prev-> val > root->val){
        t2 = root;
        if(t1 == NULL)
            t1 = prev;
    }

    prev = root;
    inorder(root->right);
}

void recoverTree(struct TreeNode* root) {
    prev = NULL;
    t1 = NULL;
    t2 = NULL;
    inorder(root);

    if(t1 && t2){
        int tmp = t1->val;
        t1->val = t2->val;
        t2->val = tmp;
    }
}

File: python/tree/recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inOrder(self,root):
        if root is None:
            return

        self.inOrder(root.left)

        if self.prev and self.prev.val > root.val:
            self.t2 = root
            if self.t1 is None:
                self.t1 = self.prev

        self.prev = root;

        self.inOrder(root.right)

    def recoverTree(self, root):
        self.prev = None
        self.t1 = None
        self.t2 = None

        self.inOrder(root)

        self.t1.val, self.t2.val = self.t2.val,self.t1.val
